### PR TITLE
feat: implement read write locks in threading and use them for CActiveMasternodeManager::cs

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented
We have some caches or other information in codebase which are read from a lot; but rarely written to. We can use a RW lock here instead of a normal Mutex

## What was done?
Implement a RW lock and use them

## How Has This Been Tested?
Hasn't been much; looking for review atm. Maybe should deploy this on testnet for a bit and make sure it doesn't break. 
## Checklist:
  _Go over all the following points, and put an `x` in all the boxes that apply._
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls check be16627d8b

File: src/sync.h [hunk 1]
```diff
@@ -123,6 +139,9 @@ using RecursiveMutex = AnnotatedMixin<std::recursive_mutex>;
 
 /** Wrapped mutex: supports waiting but not recursive locking */
 typedef AnnotatedMixin<std::mutex> Mutex;
+/** Wrapped shared mutex: supports read locking via .shared_lock, exlusive locking via .lock;
+ * does not support recursive locking */
+typedef SharedAnnotatedMixin<std::shared_mutex> SharedMutex;
```
Comment (knst) [inline src/sync.h:144]: nit: why won't use modern syntax such as:
`using SharedMutex = SharedAnnotatedMixin<std::shared_mutex>`

File: src/sync.h [hunk 2]
```diff
@@ -223,16 +242,77 @@ class SCOPED_LOCKABLE UniqueLock : public Base
...
+#endif
+            Base::lock();
+#ifdef DEBUG_LOCKCONTENTION
+        }
+#endif
+    }
+
+    bool TrySharedEnter(const char* pszName, const char* pszFile, int nLine)
+    {
+        EnterCritical(pszName, pszFile, nLine, Base::mutex(), true);
```
Comment (knst) [inline src/sync.h:264]:
```suggestion
        EnterCritical(pszName, pszFile, nLine, Base::mutex(), /*fTry = */ true);
```

File: src/sync.h [hunk 3]
```diff
@@ -223,16 +242,77 @@ class SCOPED_LOCKABLE UniqueLock : public Base
...
+    {
+        if (!pmutexIn) return;
+
+        *static_cast<Base*>(this) = Base(*pmutexIn, std::defer_lock);
+        if (fTry) {
+            TrySharedEnter(pszName, pszFile, nLine);
+        } else {
+            SharedEnter(pszName, pszFile, nLine);
+        }
+    }
```
Comment (knst) [inline src/sync.h:292]: <s> why we need  even to consider a situation when `pmutexIn` can be non-null ptr? </s>
ok it seems as copy-paste of `UniqueLock`

```suggestion
```

File: src/evo/mnauth.cpp [hunk 4]
```diff
@@ -24,38 +24,36 @@ void CMNAuth::PushMNAUTH(CNode& peer, CConnman& connman, const CBlockIndex* tip)
...
+    //   node1 -> Eve -> node2
+    // This is ok as we only use MNAUTH as a DoS protection and not for sensitive stuff
+    int nOurNodeVersion{PROTOCOL_VERSION};
+    if (Params().NetworkIDString() != CBaseChainParams::MAIN && gArgs.IsArgSet("-pushversion")) {
+        nOurNodeVersion = gArgs.GetArg("-pushversion", PROTOCOL_VERSION);
+    }
+    const bool is_basic_scheme_active{DeploymentActiveAfter(tip, Params().GetConsensus(), Consensus::DEPLOYMENT_V19)};
+    auto pk = ::activeMasternodeManager->GetPubKey();
+    const CBLSPublicKeyVersionWrapper pubKey(pk, !is_basic_scheme_active);
+    uint256 signHash = [&]() {
```
Comment (knst) [inline src/evo/mnauth.cpp:48]: missing `const`:

```suggestion
    const uint256 signHash = [&]() {
```

**Review by knst — APPROVED** [review]
LGTM

**Comment by knst** [comment]
hm, what if there would be too many reading operations which are running in many threads - is there any chance that regular lock ever be acquired? or it would wait until all reading operations are done?

**Comment by PastaPastaPasta** [comment]
> hm, what if there would be too many reading operations which are running in many threads - is there any chance that regular lock ever be acquired? or it would wait until all reading operations are done?

From the [paper](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#shared_mutex_imp) that introduces them


> shared_mutex Reference Implementation

> shared_mutex can certainly be implemented on top of an OS supplied read-write mutex. However a portable subset of the implementation is shown here for the purpose of motivating the existence of cond_var: the razor-thin layer over the OS condition variable.

> A secondary motivation is to explain the lack of reader-writer priority policies in shared_mutex. This is due to an algorithm credited to Alexander Terekhov which lets the OS decide which thread is the next to get the lock without caring whether a unique lock or shared lock is being sought. This results in a complete lack of reader or writer starvation. It is simply fair.

> Below is most of the implementation of shared_mutex demonstrating that with this proposal, very high quality synchronization devices can be easily coded, and with the same efficiency as if they were coded straight to the OS. The timed-locking functions are omitted only for brevity. They are similar to the locking functions but make use of cond_var::timed_wait. Dependence on the lower-level C++ API is highlighted.

Based on this I don't think there is an ability for starvation

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by knst — APPROVED** [review]
utACK 069282611c5058d0ea78ca83f9709034fe2c5ffd

**Pull request merged**